### ISSUE TITLE
Angle Assessment as a Function of Object Distance

### ISSUE THREAD

**Issue by Xexxar** [ppy/osu-performance#68]
Hello,

**The Problem**
I probably don't even need to include this section but... It is well known that one of the largest issues with the PP algorithm is the lack of angles being factored in the difficulty of a map. Due to this flaw, many maps are either underweighted or overweighted. Thus, a new method of assessing this type of difficulty is required.

**Required Changes to the Algorithm**
Beyond alterations to diffcalc itself, the introduction of a method that returns the acute angle formed between the previous note and the note following the object in question is required. With this method, creating new tools for balancing patterns with special attention to angles can created. Effectively what would need to be done is the following:

```
            int x1 = prev.X - current.X;
            int y1 = prev.Y - current.Y;
            int x2 = next.X - current.X;
            int y2 = next.Y - current.Y;
            int dotProduct = x1 * x2 + y1 * y2;
            int determinant = x1 * y2 - y1 * x2;
            double angle = Math.Atan2(determinant, dotProduct);
```

**Why Distance is Preferred**
In trying to assess angles, it is hard to make general rules without assessing the absolute distance between hitobjects. The reason for this is simple. While it is true that jumps which are linear in nature and typically harder than back and forth jumps, the same is not true for streams, which are typically easier if they are in a straight line. Thus, I propose separating angle buffs for speed and aim which depend on functions of distance. While it may be an over generalization, jumps tend to be more difficult the more linear they are, where as streams are more difficult the more harshly "curved" or non linear they are (smaller interior angles). Some people may argue that time elapsed is better, but as it is hard to differential a 280 bpm alt map with 280 bpm jumps, I do not think this method will work well. Thus, let's go into the specifics of how this might work.

**Speed Case**
As speed is both intended to balance single tapping speed and stream speed, it can be very difficult to balance the two in a single algorithm without adversely affecting both parts. Thus, what I want to do is introduce scaling based off the angle of the object. A stream of angle 135 or less should start to see a buff for being more difficult to play, up until 90 degrees. Here is what that code would look like.

```protected override double StrainValueOf(OsuDifficultyHitObject current)
{
double distance = current.TravelDistance + current.JumpDistance;
double angle = current.(NewAngleMethod)
double angleBonus = 1.0;

if (angle<135)
angleBonus = (135 - angle / 45) * .25 + 1.;
else if (angle=<90)
angleBonus = 1.25;

double speedValue;
            if (distance > single_spacing_threshold)
                speedValue = 2.0;
            else if (distance > stream_spacing_threshold)
                speedValue = 1.6 + 0.4 * (distance - stream_spacing_threshold) / (single_spacing_threshold - stream_spacing_threshold);
            else if (distance > almost_diameter)
                speedValue = 1.2 + 0.4 * (distance - almost_diameter) / (stream_spacing_threshold - almost_diameter);
            else if (distance > almost_diameter / 2)
                speedValue = 0.95 + 0.25 * (distance - almost_diameter / 2) / (almost_diameter / 2);
            else
                speedValue = 0.95;

            return speedValue * angleBonus / current.StrainTime;
}
```
Which can be visualized with the following image.
![](http://puu.sh/CdEpy/bf0386774e.png)
Please observe that the `speedValue`s for distance were altered with this proposal. In order to maintain consistent single tapping stamina speed values, we can now lower the extreme scaling that occurs from `stream_spacing_threshold` and `single_spacing_threshold`, because most single taps will receive the 1.25x buff that has been added. Thus, a single tap would receive it's standard `2.0x1.25 = 2.5` speed value, while linear streams no longer get the extreme scaling in between the two thresholds that used to exist. This change alone will GREATLY impact the overweighted stream maps that abuse this oversight in the algorithm (This may have actually been my main goal with this proposal) (These values are placeholder, testing will need to be done before anything is finalized).

**Aim Case**
Utilizing a similar idea as before, we can use this new angle method to rewards wide angle jumps like the ones seen in older maps and other "anti-pp" maps. In this case however, in order to not rewards aim on streams since they're linear, only buffing the portion of distance that exceeds regular streaming distance seems to be the best solution. For example, buffing distance that exceeds the `single_spacing_threshold` distance. While it is arbitrary, here is an example of how such a buff might work for angles greater than 75.
![](http://puu.sh/CdEVH/0add4d6baf.png)
And here is how the code for such a suggestion may be implemented.

```
protected override double StrainValueOf(OsuDifficultyHitObject current)
{
           double angle = current.(NewAngleMethod)
           double angleBonus = 1.0
           if (current.JumpDistance > single_spacing_threshold)
                 {
                 if (angle>75)
                        angleBonus = (-75 + angle / 45) / 2;
                 else if (angle > 120)
                        angleBonus = .5;
                return angleBonus * Math.Pow(current.JumpDistance - single_spacing_threshold, 0.99) + Math.Pow(current.TravelDistance, 0.99) + Math.Pow(current.JumpDistance, 0.99)) / current.StrainTime;
                 }
          else
          return Math.Pow(current.TravelDistance, 0.99) + Math.Pow(current.JumpDistance, 0.99)) / current.StrainTime;
}
```
Thus scaling any distance greater than `single_spacing_threshold` by the `angleBonus` (assuming my programming is correct). This would reward angles that exceed 75 with a scale that increases up to .5 for angles that are greater than 120 degrees.

**Conclusion**
While this solution still does not take into account many factors of angle calculation such as flow, flow changes, weird patterns etc, it is a step in the right direction towards remedying major issues within the algorithm. While I didn't mention it in the initial introduction, the most important change from this will be the rebalancing of streams that exceed the `stream_spacing_threshold` by nerfing the 1.6 -> 2.5 scaling that occurs between the distances of 110 and 125. I hope some test versions of this suggestion can be implemented so we can see what this might be able to do to improve the algorithm. 

As always, thank you for reading and have a wonderful day.

**Comment by Feuerholz** [comment]
hype

Might be better to base the decision of whether to buff or nerf low angles on speed (as in time between notes) rather than distance though, the actual difference in difficulty comes from flow aim vs. snap aim, and whether a player is able to snap depends much more heavily on speed than on distance (see: atomosphere triples easier than triangles, dopamine triangles hard even though low spacing and would be super easy if they were the same distance in a straight line)

**Comment by diamondburned** [comment]
Working on something to calculate atm, I have a question. Do I account angle a) or b) in the pic: https://ptpb.pw/IVwE.png

I'm assuming it's a) since it doesn't make sense to determine an acute angle based on a 360 degrees plane

**Comment by diamondburned** [comment]
For this part: 

```cs
protected override double StrainValueOf(OsuDifficultyHitObject current)
{
           double angle = current.(NewAngleMethod)
           double angleBonus = 1.0
           if (current.JumpDistance > single_spacing_threshold)
                 {
                 if (angle>75)
                        angleBonus = (-75 + angle / 45) / 2;
                 else if (angle > 120)
                        angleBonus = .5;
                return angleBonus * Math.Pow(current.JumpDistance - single_spacing_threshold, 0.99) + Math.Pow(current.TravelDistance, 0.99) + Math.Pow(current.JumpDistance, 0.99)) / current.StrainTime;
                 }
          else
          return Math.Pow(current.TravelDistance, 0.99) + Math.Pow(current.JumpDistance, 0.99)) / current.StrainTime;
}
```

Do you mean: 

```cs
protected override double StrainValueOf(OsuDifficultyHitObject current)
{
           double angle = current.(NewAngleMethod)
           double angleBonus = 1.0
           if (current.JumpDistance > single_spacing_threshold)
                 {
                 if (angle>75)
                        angleBonus = (-75 + angle / 45) / 2;
                 else if (angle > 120)
                        angleBonus = .5;
                 }
                 return angleBonus * Math.Pow(current.JumpDistance - single_spacing_threshold, 0.99) + Math.Pow(current.TravelDistance, 0.99) + Math.Pow(current.JumpDistance, 0.99)) / current.StrainTime;
          else
          return Math.Pow(current.TravelDistance, 0.99) + Math.Pow(current.JumpDistance, 0.99)) / current.StrainTime;
}``` 

(changes at L11-12)

**Comment by smoogipoo** [comment]
You're confused about TravelDistance and JumpDistance because you're not using osu!lazer to calculate this.

If you're wanting to contribute, please do so using osu!, where you'll find said variables referenced in places such as https://github.com/ppy/osu/blob/master/osu.Game.Rulesets.Osu/Difficulty/Skills/Speed.cs

**Comment by diamondburned** [comment]
I'm still unsure about contributing, I'm working on a calculator to determine this first. I'm starting with oppai, but already stumbling upon issues.

**Comment by smoogipoo** [comment]
Figure it out on your own then, I have provided enough information in my comment to do so. This is not a tutorial on how to build your own difficulty calculator unless it is related to osu!lazer.

**Comment by visualer** [comment]
Redditors do have some good points. Thresholds may not be used to avoid abrupt changes, and any function used should be smooth, or at least differentiable to the 2nd or 3rd order.

**Comment by Tom94** [comment]
Differentiability is nice, but I feel a continuous function that is a reasonable approximation of something differentiable is already sufficient in many cases (think piecewise-linear approximations with a couple vertices). The error of approximating some function as piecewise-linear is O(n^-2) where n is the number of vertices, after all.

All that said, it would probably not be a bad idea to implement and use B-splines of order 2 or 3 for everything involving thresholds in case somebody feels up to the task.

**Comment by smoogipoo** [comment]
Got a few sample values for `single_spacing_threshold` and `stream_spacing_threshold` that you're interested in testing out? Will deploy a test by this weekend with them.

**Comment by diamondburned** [comment]
> Figure it out on your own then, I have provided enough information in my comment to do so. This is not a tutorial on how to build your own difficulty calculator unless it is related to osu!lazer.

Then what about my suggestions? Is it going to be ignored just because I added a small detail?

**Comment by Xexxar** [comment]
smoogipoo: I'm referencing `single_spacing_threshold` and `stream_spacing_threshold` as their pre-existing forms within speed.cs

```        
private const double single_spacing_threshold = 125;
private const double stream_spacing_threshold = 110;
```

**Comment by smoogipoo** [comment]
Here's some sample data with this change:

Original (reference, no changes):  
http://original.smgi.me

Angles (this change only, no speed bonus/length bonus/etc):
http://xexxar-angles.smgi.me

**Comment by diamondburned** [comment]
2 days, still no answers, and I got told to get out...

**Comment by diamondburned** [comment]
Found another issue on the Aim strain part: 

```cs
protected override double StrainValueOf(OsuDifficultyHitObject current)
{
           double angle = current.(NewAngleMethod)
           double angleBonus = 1.0
           if (current.JumpDistance > single_spacing_threshold)
                 {
                 if (angle>75)
                        angleBonus = (-75 + angle / 45) / 2;
                 else if (angle > 120)
                        angleBonus = .5;
                return angleBonus * Math.Pow(current.JumpDistance - single_spacing_threshold, 0.99) + Math.Pow(current.TravelDistance, 0.99) + Math.Pow(current.JumpDistance, 0.99)) / current.StrainTime;
                 }
          else
          return Math.Pow(current.TravelDistance, 0.99) + Math.Pow(current.JumpDistance, 0.99)) / current.StrainTime;
}```

Quite stuck on the first return part: 

```cs
angleBonus * Math.Pow(current.JumpDistance - single_spacing_threshold, 0.99) + Math.Pow(current.TravelDistance, 0.99) + Math.Pow(current.JumpDistance, 0.99))``` is missing an opening `(`. Where is this placed? Is it in front of angleBonus or after angleBonus?

**Comment by smoogipoo** [comment]
Fwiw I believe my implementation of these changes to be correct, you can refer to them from https://github.com/ppy/osu/compare/master...smoogipoo:xexxar-angles . Should answer any queries you have about how the code should work. (The opening `(` can only be around everything otherwise aim would spiral out of control)

**Comment by girantinas** [comment]
This line in Xexxar's code:
`angleBonus = (-75 + angle / 45) / 2;` is a typo and should read:
`angleBonus = (-75 + angle)/90` which is the same as what smoogipoo has put, albeit he coded it into radian mode.

**Comment by smoogipoo** [comment]
Have recalculated scores in http://xexxar-angles.smgi.me to reflect the proposed changes from Discord.

**Comment by diamondburned** [comment]
Just in time for me to complete mine, all up to date! Thanks Francesco for clearing up some of the troubles.

Source: https://gitlab.com/diamondburned/goppai/tree/sexcar_angle_rework/examples/top100 (Linux users should be able to easily build this)

Example top play calculator: https://ptpb.pw/7ObF.exe (CLI application, run from command line)

**Comment by PowerChaos125** [comment]
the coding logic of angle bonus for aim is incorrect

```cs
                 if (angle>75)
                        angleBonus = (-75 + angle / 45) / 2;
                 else if (angle > 120)
                        angleBonus = .5;
```

The second if statement can never be reached, because there is no case of `angle` that can satisfy the second condition while not satisfying the first condition.
I think you meant to do this:

```cs
                 if (angle > 120)
                        angleBonus = .5;
                 else if (angle > 75)
                        angleBonus = (-75 + angle / 45) / 2;
```

**Comment by PowerChaos125** [comment]
Also, I don't think we need angle bonus for aim at all. The argument for that wide jump is harder than 1-2 jump is subjective. This largely depends on player's reference. Personally, I think wide jump and back-and-forth jump are the same.

Even if needed so, I think the scaling range of angle for aim is too small. Some thing like 45-135 would work better than 75-120

**Comment by diamondburned** [comment]
> Also, I don't think we need angle bonus for aim at all. The argument for that wide jump is harder than 1-2 jump is subjective. This largely depends on player's reference. Personally, I think wide jump and back-and-forth jump are the same.
> 
> Even if needed so, I think the scaling range of angle for aim is too small. Some thing like 45-135 would work better than 75-120

The proposal above you took from is really old, and I agree that he should change it. As of now, the only way to know the latest change is to check https://github.com/ppy/osu/pull/3839 and this comment section.

**Comment by diamondburned** [comment]
@smoogipoo Can I know why this `math.Min` is here? https://github.com/ppy/osu/compare/master...smoogipoo:xexxar-angles#diff-b53689abe7de64ce5830e35492517d24R25

I've noticed that it keeps the Speed pp consistent between my trials, but the Aim pp doesn't have t hat, leading to an unreasonable-looking result.

**Comment by smoogipoo** [comment]
@diamondburned Because in the piecewise version it was essentially clamped to that value.

**Comment by Karthyz** [comment]
> It is worth pointing out that the difficulty of linear jumps is also quite considerably affected by the change of spacing between each circle, for example if a low angle linear jump of 4 circles suddenly lowers and increases spacing between each circle, the flow of the jump will be affected, and it will require more effort than just swiping your aiming hand at a consistent pace.

I think this would start getting more into reading than aim. It starts getting alot more complex because some patterns that have this where certain notes in a linear pattern have different spacing actually end up being easier to read and aim than just an ordinary linear pattern. It can also have the opposite affect, but it ends up getting very complex and I think it's probably outside of the scope of this change.

**Comment by VINXIS** [comment]
Wide angles are possibly not giving enough of a bonus to offset the 2.5->1.95 jump speed bonus. A prime example of a map that has a TON of wide angles, but fails to see any bonus is [Pluto](https://xexxar-angles.smgi.me/beatmapsets/45074#osu/146957). The map's #1 play was originally 393pp, but after the bonus, the gain Only raises it to 394pp, which seems like a rediculously low bonus for a wide angle buff. The nomod plays somehow also seem to have a decrease in PP as well.

**Comment by smoogipoo** [comment]
Closing this as it has been implemented now.

**Comment by tr3sleches** [comment]
Dear god why was this implemented.
